### PR TITLE
Begin adding type annotations

### PR DESCRIPTION
**Pull request by peastman**
So far I've only added annotations to KerasModel, plus fixing some errors identified by mypy in other files.  I want to add them to a few other core classes in this PR, and then we can gradually add them to the rest of the code.

It's going to take us a while to figure out best practices for how to identify types.  For example, there are a lot of places where the docstring says an argument should be a list, but actually a tuple works fine too, so we need to decide when to insist on `List` as the type and when to say `Sequence` to allow either one.  Numpy arrays are another ambiguous case.  I've followed the principle that when we're certain an output value is really an array it should be listed as `np.ndarray`, but input values are usually listed as `Sequence` to accept array-like objects such as lists of numbers.

Then there are cases where giving a precise type just seems like more trouble than it's worth.  The `generator` argument to `fit_generator()` is supposed to be a generator (but really anything iterable is fine) that produces tuples of (inputs, labels, weights), where each of those is a list (but in practice any sequence is fine) of Numpy arrays or array-like objects or tensors.  It's possible to exactly describe that, but instead I just gave the type as `Iterable[Tuple[Any, Any, Any]]`.  Maybe I should change that to `Iterable[Tuple[Sequence, Sequence, Sequence]]`?

### CODE REVIEW

File: deepchem/models/keras_model.py [hunk 1]
```diff
@@ -4,19 +4,26 @@
...
+from deepchem.metrics import Metric
 from deepchem.models.losses import Loss
 from deepchem.models.models import Model
-from deepchem.models.optimizers import Adam
-from deepchem.trans import undo_transforms
+from deepchem.models.optimizers import Adam, Optimizer, LearningRateSchedule
+from deepchem.trans import Transformer, undo_transforms
 from deepchem.utils.evaluate import GeneratorEvaluator
 
+from typing import Any, Callable, Dict, Iterable, List, Optional, Sequence, Tuple, TypeVar, Union
```
Comment (rbharath) [inline deepchem/models/keras_model.py:21]: Should we add a `deepchem.types` module to gather all DeepChem types in one place?

Comment (ncfrey) [reply deepchem/models/keras_model.py:21]: I think a `deepchem.types` module and/or some dedicated documentation explaining what the DeepChem type conventions are would be really helpful. Type annotations are great, but it's really easy to become inconsistent across modules. The `mypy` [cheatsheet](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html) might be a good starting point.

Comment (peastman) [reply deepchem/models/keras_model.py:21]: Good idea.  How about `deepchem.utils.typing`?  I've already come across a couple of compound types that are likely to appear in a lot of places.

Comment (rbharath) [reply deepchem/models/keras_model.py:21]: Works great! It might also be worth adding a `docs/typing.rst` docs file that lays out DeepChem type conventions for newcomers

**Comment by peastman** [comment]
Here's an interesting case I came across in `DiskDataset.write_data_to_disk()`.

```python
    if X is not None:
      out_X = "%s-X.npy" % basename
      save_to_disk(X, os.path.join(data_dir, out_X))
    else:
      out_X = None
```

mypy objects to the last line: it inferred from the first assignment that `out_X` has type `str`, so assigning `None` to it is an error.  I can add an annotation to clarify that the actual type is `Optional[str]`, but then it reports an error in the call to `join()`, since the arguments can't be `None`.  I could work around that by first adding a check `if out_X is not None`, but that's just silly.  Obviously it isn't `None`.  I just assigned it in the previous line!  Or instead I could add a `type: ignore` comment to that line, but that feels like cheating.  Is there a recommended way of handling this case?

**Comment by peastman** [comment]
MultitaskFitTransformRegressor takes an argument called `fit_transformers`.  The docstring says this should be a "List of dc.trans.FitTransformer objects".  There's no such class.  It assumes the provided objects have a method called `X_transform()`.  Only two transformers have methods with that name, and it's not clear to me that they actually do the same thing: CoulombFitTransformer and IRVTransformer.

What is the intended type for this argument, and how is it supposed to work?

**Comment by rbharath** [comment]
Hmm, that's a good point. Here's a usage example for `MultitaskFitTransformRegressor`: https://github.com/deepchem/deepchem/blob/master/deepchem/trans/transformers.py#L1136

I think we can set the type to `Transformer` and apply a fix to `MultitaskFitTransformRegressor` at the line https://github.com/deepchem/deepchem/blob/master/deepchem/models/fcnet.py#L391. We could replace this line with
```
assert transformer.transform_X and not (transformer.transform_y or transformer.transform_w)
X_b = transformer.transform_arrays(X_b, None, None)
```
Then we'd need to fix the API for `CoulombFitTransformer` and `IRVTransformer` to match the API for the other transformers

**Comment by peastman** [comment]
That works, thanks!

File: deepchem/models/keras_model.py (hunk 1 above)
Comment (peastman) [reply deepchem/models/keras_model.py]: What about a more general `coding.rst` file for coding conventions?  In addition to type annotations, it could also discuss unit tests, doc strings, code formatting, etc.

Comment (rbharath) [reply deepchem/models/keras_model.py]: Sounds like a great idea!

**Comment by peastman** [comment]
This is ready for review.

File: deepchem/data/datasets.py [hunk 2]
```diff
@@ -374,7 +380,8 @@ def itersamples(self):
     """
     raise NotImplementedError()
 
-  def transform(self, fn, **args):
+  def transform(self, fn: Callable[[np.ndarray, np.ndarray, np.ndarray], Tuple[
```
Comment (rbharath) [inline deepchem/data/datasets.py:383]: Could we introduce an auxiliary type here? Say

```
TransformerBatch = Tuple[np.ndarray, np.ndarray, np.ndarray]
```
Then we write something like
```
Callable[TransformerBatch, TransformerBatch]
```

File: deepchem/data/datasets.py [hunk 3]
```diff
@@ -491,7 +499,7 @@ def gen_data():
 
     return tf.data.Dataset.from_generator(gen_data, dtypes, shapes)
 
-  def make_pytorch_dataset(self, epochs=1, deterministic=False):
+  def make_pytorch_dataset(self, epochs: int = 1, deterministic: bool = False):
```
Comment (rbharath) [inline deepchem/data/datasets.py:502]: Should we add a return type here?

File: deepchem/data/datasets.py [hunk 4]
```diff
@@ -444,13 +452,13 @@ def get_statistics(self, X_stats=True, y_stats=True):
...
 
   def make_tf_dataset(self,
-                      batch_size=100,
-                      epochs=1,
-                      deterministic=False,
-                      pad_batches=False):
+                      batch_size: int = 100,
+                      epochs: int = 1,
+                      deterministic: bool = False,
+                      pad_batches: bool = False):
```
Comment (rbharath) [inline deepchem/data/datasets.py:461]: Should we add a return type here?

File: deepchem/data/datasets.py [hunk 5]
```diff
@@ -876,7 +896,7 @@ def __iter__(self):
     return TorchDataset()
 
   @staticmethod
-  def from_DiskDataset(ds):
+  def from_DiskDataset(ds: "DiskDataset") -> "NumpyDataset":
```
Comment (rbharath) [inline deepchem/data/datasets.py:899]: Why is this "DiskDataset" instead of `DiskDataset` directly? Sorry if this is a newbie question about mypy!

File: docs/coding.rst [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+avoids disagreements about how a piece of code should be formatted.
+
+Whenever you modify a file, run :code:`yapf` on it to reformat it before
+checking it in.
+
+.. code-block:: bash
+
+  yapf -i <modified file>
+
+Yapf is run on every pull request to make sure the formatting is correct, so if
```
Comment (rbharath) [inline docs/coding.rst:21]: Could you add a line about making sure to use the same version of `yapf` as DeepChem is currently using? We're on 0.22, and there seem to be big differences between the latest 0.30 and 0.22

File: docs/coding.rst [hunk 7]
```diff
@@ -0,0 +1,100 @@
+Coding Conventions
```
Comment (rbharath) [inline docs/coding.rst:1]: This file is great! Will help a lot in onboarding new developers :)

File: deepchem/data/datasets.py (hunk 3 above)
Comment (peastman) [reply deepchem/data/datasets.py:502]: We can't, because the return type is a PyTorch class and we don't want to `import torch` outside the function.

File: deepchem/data/datasets.py (hunk 4 above)
Comment (peastman) [reply deepchem/data/datasets.py:461]: Same here.  Declaring the return type would require importing TensorFlow at the top level.

File: deepchem/data/datasets.py (hunk 5 above)
Comment (peastman) [reply deepchem/data/datasets.py:899]: This is a workaround for the fact that Python doesn't allow forward references.  We can't refer to DiskDataset here, because it hasn't been defined yet at this point in the file.  The workaround is that you put it in quotes, and mypy evaluate the string.

File: docs/coding.rst (hunk 6 above)
Comment (peastman) [reply docs/coding.rst:21]: Good idea!  I added it.

File: deepchem/data/datasets.py (hunk 2 above)
Comment (peastman) [reply deepchem/data/datasets.py:383]: The argument to `fn` isn't a tuple, it's just three separate arrays.  I wonder if we should change this method to take a Transformer object instead?  I'm investigating this as part of #1990.

File: deepchem/data/datasets.py (hunk 5 above)
Comment (rbharath) [reply deepchem/data/datasets.py:899]: Ah that makes sense!

File: deepchem/data/datasets.py (hunk 2 above)
Comment (rbharath) [reply deepchem/data/datasets.py:383]: Ah that's interesting. Only transformers use the `Dataset.transform` method in any case so might be worth tying them together if it helps clean things up

**Comment by peastman** [comment]
IRVTransformer doesn't extend Transformer???  Is that intentional or an oversight?

**Comment by rbharath** [comment]
Definitely looks like an unfortunate oversight!

**Comment by peastman** [comment]
Tests are passing now.

**Comment by rbharath** [comment]
Looks great! I'm excited to get this in :)
